hpx/runtime/serialization/detail/polymorphic_centralized_factory.hpp

```diff
@@ -0,0 +1,218 @@
+//  Copyright (c) 2015 Anton Bikineev
+//  Copyright (c) 2014 Thomas Heller
+//
+//  Distributed under the Boost Software License, Version 1.0.
+//  See accompanying file LICENSE_1_0.txt or copy at
+//  http://www.boost.org/LICENSE_1_0.txt)
+
+#ifndef HPX_SERIALIZATION_POLYMORPHIC_CENTRALIZED_FACTORY_HPP
+#define HPX_SERIALIZATION_POLYMORPHIC_CENTRALIZED_FACTORY_HPP
+
+#include <hpx/runtime/serialization/serialization_fwd.hpp>
+#include <hpx/runtime/serialization/detail/polymorphic_intrusive_factory.hpp>
+#include <hpx/traits/polymorphic_traits.hpp>
+#include <hpx/util/static.hpp>
+#include <hpx/util/safe_lexical_cast.hpp>
+#include <hpx/exception.hpp>
+
+#include <boost/preprocessor/stringize.hpp>
+#include <boost/unordered_map.hpp>
+#include <boost/atomic.hpp>
+#include <boost/mpl/bool.hpp>
+
+#include <map>
+
+#include <hpx/config/warnings_prefix.hpp>
+
+namespace hpx { namespace serialization { namespace detail
+{
+    struct type_name_desc_type
+    {
+        BOOST_STATIC_CONSTEXPR boost::uint32_t invalid = ~0u;
+
+        type_name_desc_type(std::string name,
+                boost::uint32_t desc = invalid)
+            : name(std::move(name)), desc(desc)
+        {}
+
+        std::string name;
+        boost::uint32_t desc;
+    };
+
+    BOOST_FORCEINLINE
+    static bool operator<(
+            const hpx::serialization::detail::type_name_desc_type& first,
+            const hpx::serialization::detail::type_name_desc_type& second)
+    {
+        return first.desc < second.desc;
+    }
+
+    class HPX_EXPORT polymorphic_centralized_factory: boost::noncopyable
+    {
+    public:
+        typedef void* (*ctor_type) ();
+        typedef std::multimap<type_name_desc_type, ctor_type> ctor_map_type;
+
+        static void add_factory_function(const std::string& name, ctor_type fun)
+        {
+            if(name.empty())
+            {
+                HPX_THROW_EXCEPTION(serialization_error
+                  , "polymorphic_intrusive_factory::register_class"
+                  , "Cannot register a factory with an empty name");
+            }
+            auto& map = polymorphic_centralized_factory::instance().map_;
```

Comment:
@K-ballo, thanks.
Do you prefer having one static instance method with non-static other methods instead of this approach (like it's done in other factories). I liked this, because it simply doesn't require user to type instance() several times